File: nosleep/nosleep.py
#!/usr/bin/env python3
import argparse
import subprocess
import sys
import time

PORTAL_DEST = "org.freedesktop.portal.Desktop"
PORTAL_PATH = "/org/freedesktop/portal/desktop"
PORTAL_IFACE = "org.freedesktop.portal.Inhibit"

# Portal flags: 1=logout, 2=user-switch, 4=suspend, 8=idle
INHIBIT_IDLE = 8


def build_gdbus_args(app_id: str, reason: str) -> list[str]:
    # gdbus expects a{sv} with variant values.
    options = "{'reason': <'" + reason.replace("'", "\\'") + "'>}"
    return [
        "gdbus",
        "call",
        "--session",
        "--dest",
        PORTAL_DEST,
        "--object-path",
        PORTAL_PATH,
        "--method",
        f"{PORTAL_IFACE}.Inhibit",
        app_id,
        "",  # No window handle
        str(INHIBIT_IDLE),
        options,
    ]


def main() -> int:
    parser = argparse.ArgumentParser(
        description="Prevent idle sleep/screensaver via xdg-desktop-portal."
    )
    parser.add_argument(
        "--app-id",
        default="com.example.nosleep",
        help="Application ID string for the portal request.",
    )
    parser.add_argument(
        "--reason",
        default="Keeping the session awake",
        help="Human-readable reason for the inhibit request.",
    )
    args = parser.parse_args()

    gdbus_args = build_gdbus_args(args.app_id, args.reason)

    try:
        result = subprocess.run(
            gdbus_args,
            check=True,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
        )
    except FileNotFoundError:
        print("gdbus not found. Install the 'glib2.0-bin' package.", file=sys.stderr)
        return 1
    except subprocess.CalledProcessError as exc:
        print("Failed to request inhibit via xdg-desktop-portal.", file=sys.stderr)
        if exc.stderr:
            print(exc.stderr.strip(), file=sys.stderr)
        return exc.returncode

    # The returned handle is tied to this process's bus name; keep it alive.
    handle = result.stdout.strip()
    print("Inhibit active. Press Ctrl+C to release.")
    if handle:
        print(f"Portal handle: {handle}")

    try:
        while True:
            time.sleep(60)
    except KeyboardInterrupt:
        print("Releasing inhibit and exiting.")
        return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: nosleep/nosleep.js
#!/usr/bin/env node
'use strict';

const { spawnSync } = require('child_process');

const PORTAL_DEST = 'org.freedesktop.portal.Desktop';
const PORTAL_PATH = '/org/freedesktop/portal/desktop';
const PORTAL_IFACE = 'org.freedesktop.portal.Inhibit';
const INHIBIT_IDLE = 8; // idle/screen saver

function parseArgs(argv) {
  const args = { appId: 'com.example.nosleep', reason: 'Keeping the session awake' };
  for (let i = 2; i < argv.length; i++) {
    const arg = argv[i];
    if (arg === '--app-id' && argv[i + 1]) {
      args.appId = argv[++i];
    } else if (arg === '--reason' && argv[i + 1]) {
      args.reason = argv[++i];
    } else if (arg === '-h' || arg === '--help') {
      args.help = true;
    }
  }
  return args;
}

function buildGdbusArgs(appId, reason) {
  const escapedReason = reason.replace(/'/g, "\\'");
  const options = `{'reason': <'${escapedReason}'>}`;
  return [
    'call',
    '--session',
    '--dest',
    PORTAL_DEST,
    '--object-path',
    PORTAL_PATH,
    '--method',
    `${PORTAL_IFACE}.Inhibit`,
    appId,
    '',
    String(INHIBIT_IDLE),
    options,
  ];
}

function main() {
  const args = parseArgs(process.argv);
  if (args.help) {
    console.log('Usage: nosleep.js [--app-id ID] [--reason TEXT]');
    return 0;
  }

  const gdbusArgs = buildGdbusArgs(args.appId, args.reason);
  const res = spawnSync('gdbus', gdbusArgs, { encoding: 'utf8' });

  if (res.error) {
    if (res.error.code === 'ENOENT') {
      console.error("gdbus not found. Install the 'glib2.0-bin' package.");
      return 1;
    }
    console.error('Failed to run gdbus.', res.error.message || res.error);
    return 1;
  }

  if (res.status !== 0) {
    console.error('Failed to request inhibit via xdg-desktop-portal.');
    if (res.stderr) {
      console.error(res.stderr.trim());
    }
    return res.status || 1;
  }

  const handle = (res.stdout || '').trim();
  console.log('Inhibit active. Press Ctrl+C to release.');
  if (handle) {
    console.log(`Portal handle: ${handle}`);
  }

  setInterval(() => {}, 60 * 1000);
  return 0;
}

process.on('SIGINT', () => {
  console.log('Releasing inhibit and exiting.');
  process.exit(0);
});

process.exitCode = main();
